### PR TITLE
Ft/batch policy eval

### PR DESCRIPTION
**Pull request by LaurenSpiegel**
Send an array of request contexts to Vault to do authorization checks as a batch.

### CODE REVIEW

**Comment by LaurenSpiegel** [comment]
@ironman-machine batch time
SCALITY_VAULT_BRANCH=ft/batchPolicyEval

**Comment by michael-zapata** [comment]
Could we have some tests to go along with it?

**Comment by LaurenSpiegel** [comment]
I added tests to integration.  We are not doing in memory policy evaluation.

File: lib/api/api.js [hunk 1]
```diff
@@ -21,19 +21,22 @@ import objectHead from './objectHead';
...
 const api = {
     callApiMethod(apiMethod, request, log, callback, locationConstraint) {
+        let sourceBucket;
+        let sourceObject;
         if (apiMethod === 'objectCopy') {
             let source =
                 querystring.unescape(request.headers['x-amz-copy-source']);
-            // If client sends the source bucket/object with a leading /,
-            // remove it
+                // If client sends the source bucket/object with a leading /,
```
Comment (michael-zapata) [inline lib/api/api.js]: Considering this applies to the code below it, why not leave the indentation as is?

File: lib/api/api.js [hunk 2]
```diff
@@ -42,47 +45,37 @@ const api = {
...
+            authorizationResults) => {
             if (err) {
                 log.trace('authentication error', { error: err });
                 return callback(err);
             }
+            // TODO: When implement batch delete, check apiMethod
+            // and handle 'Deny' on key by key basis
+            if (authorizationResults) {
+                for (let i = 0; i < authorizationResults.length; i ++) {
+                    if (authorizationResults[i].verdict === 'Deny') {
```
Comment (michael-zapata) [inline lib/api/api.js]: Why use a string? Can't a boolean work just as well while limiting typos?

File: lib/api/api.js [hunk 3]
```diff
@@ -42,47 +45,37 @@ const api = {
...
+        return auth.doAuth(request, log, (err, userInfo,
+            authorizationResults) => {
             if (err) {
                 log.trace('authentication error', { error: err });
                 return callback(err);
             }
+            // TODO: When implement batch delete, check apiMethod
+            // and handle 'Deny' on key by key basis
+            if (authorizationResults) {
+                for (let i = 0; i < authorizationResults.length; i ++) {
```
Comment (MathieuCassagne) [inline lib/api/api.js]: You can remove the space in ```i ++``` :)

File: lib/api/api.js (hunk 1 above)
Comment (LaurenSpiegel) [inline lib/api/api.js]: fixed

File: lib/api/api.js (hunk 2 above)
Comment (LaurenSpiegel) [inline lib/api/api.js]: changed to isAllowed

File: lib/api/api.js (hunk 3 above)
Comment (LaurenSpiegel) [inline lib/api/api.js]: changed

**Comment by LaurenSpiegel** [comment]
@ironman-machine get it on
SCALITY_VAULT_BRANCH=ft/batchPolicyEval
SCALITY_INTEGRATION_BRANCH=test/multipleRequestContexts

File: lib/api/api.js [hunk 4]
```diff
@@ -42,47 +45,37 @@ const api = {
...
             if (err) {
                 log.trace('authentication error', { error: err });
                 return callback(err);
             }
+            // TODO: When implement batch delete, check apiMethod
+            // and handle 'Deny' on key by key basis
+            if (authorizationResults) {
+                for (let i = 0; i < authorizationResults.length; i++) {
+                    if (!authorizationResults[i].isAllowed) {
+                        log.trace('authorization denial from Vault');
```
Comment (rahulreddy) [inline lib/api/api.js]: It would be helpful to log which resource was denied authorization, do we have the info here?

Comment (LaurenSpiegel) [inline lib/api/api.js]: For now only object copy has more than one requestContext and the resource is the same for both checks.  I have added to the TODO note for batch delete to add that log.

**Comment by LaurenSpiegel** [comment]
End to end passed!

**Comment by rahulreddy** [comment]
👍

**Pull request merged**